Support excluding filter on mapper scan feature

apdya opened the pull request:
fix issue #548. add field  `ComponentScan.Filter[] excludeFilters()` to `@MapperScan`, so we can use `ComponentScan.Filter[]` to customized mapper excludeFilters.

hazendaz commented:
Few issues I see here
- Add code into deprecated code
- Adding aspectj but doesn't use it at all
- Alignment issues (ie adding properties that are against the next line javadocs)

@kazuki43zoo Will need to check this out but please cleanup items noted and make a case for why deprecated code is being filled out while deprecation was left.  Not seeing that was even needed off hand but didn't look deeper than that.

apdya commented:
> Few issues I see here
> 
> * Add code into deprecated code
> * Adding aspectj but doesn't use it at all
> * Alignment issues (ie adding properties that are against the next line javadocs)
> 
> @kazuki43zoo Will need to check this out but please cleanup items noted and make a case for why deprecated code is being filled out while deprecation was left. Not seeing that was even needed off hand but didn't look deeper than that


> Few issues I see here
> 
> * Add code into deprecated code
> * Adding aspectj but doesn't use it at all
> * Alignment issues (ie adding properties that are against the next line javadocs)
> 
> @kazuki43zoo Will need to check this out but please cleanup items noted and make a case for why deprecated code is being filled out while deprecation was left. Not seeing that was even needed off hand but didn't look deeper than that.

thanks for replying

- issue 1 and 3: I will modify my code and update this pr.
- issue 2: `ComponentScan.Filter[]` is from spring-context, it supports ASPECTJ FilterType, which need dependency on aspectjweaver, and this dependency is set as optional in spring-context, so this dependency need be added.

hazendaz commented on src/main/java/org/mybatis/spring/annotation/MapperScannerRegistrar.java:
```diff
@@ -31,10 +32,13 @@
 import org.springframework.core.annotation.AnnotationAttributes;
 import org.springframework.core.io.ResourceLoader;
 import org.springframework.core.type.AnnotationMetadata;
+import org.springframework.core.type.filter.*;
```
Please do not use Star imports, ass appropriate import.

hazendaz commented on src/main/java/org/mybatis/spring/annotation/MapperScannerRegistrar.java:
```diff
@@ -54,15 +58,13 @@
  */
 public class MapperScannerRegistrar implements ImportBeanDefinitionRegistrar, ResourceLoaderAware {
 
+  private ResourceLoader resourceLoader;
```
Please add a blank line between this and the next comment.

hazendaz commented on pom.xml:
```diff
@@ -316,6 +316,12 @@
       <version>6.0.0</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
```
While this marked optional, is it really required to compile here?  There isn't a single import on it.  So do we really need to state that?  IMO Aspectj is very old and interceptors replaced it a long time ago.  I rarely see code using it and manage some 2k repos in my day job and I can single digit count the ones that have aspectj.

apdya replied:
The new property `ComponentScan.Filter[]` in `@MapperScan` is reuse from Spring, and it support `FilterType.ASPECTJ` by using `org.springframework.core.type.filter.AspectJTypeFilter `, and this Class dependence on aspectjweaver, but  aspectjweaver is setted as optional dependency in spring-core, so I added this dependency, otherwise I can not use `org.springframework.core.type.filter.AspectJTypeFilter`.

hazendaz commented:
thanks for updates, few more comments.

apdya commented:
thanks for review
I have modified my code as your comment, and explain the reason to import aspectjweaver in that conversation.

hazendaz commented:
@kazuki43zoo Can you take a look at this one?

hazendaz commented:
@kazuki43zoo Can you take a look at this one?

hazendaz commented:
@apdya Tests don't build, can you please take a look?

apdya commented:
> @apdya Tests don't build, can you please take a look?

sorry I haven't run all test in my local project before commit. I update the new added test code just now, and all test passed.

kazuki43zoo commented:
> Can you take a look at this one?

Just wait a few time!

apdya commented:
> > Can you take a look at this one?
> 
> Just wait a few time!

hello, this problem has been resolved after lasted commit.

kazuki43zoo commented on src/main/java/org/mybatis/spring/annotation/MapperScannerRegistrar.java:
```diff
@@ -145,6 +157,48 @@ void registerBeanDefinitions(AnnotationMetadata annoMeta, AnnotationAttributes a
+          typeFilters.add(BeanUtils.instantiateClass(filterClass, TypeFilter.class));
+          break;
+        default:
+          throw new IllegalArgumentException("Unknown filter type " + filterType);
```
I prefer such as following message.

* `Cannot specify the 'value' or 'type' attribute if use the '{filterType}' in exclude filter of @MapperScan.`

kazuki43zoo commented on src/main/java/org/mybatis/spring/mapper/MapperScannerConfigurer.java:
```diff
@@ -191,6 +195,18 @@ public void setMarkerInterface(Class<?> superClass) {
+   *
+   * @param excludeFilters
+   *          list of TypeFilter
+   */
```
Please add `@since 3.0.3`

kazuki43zoo reviewed (changes requested):
I added some comments. 

@apdya Could you support this enhancement for the `<mybatis:scan>`(XML namespace feature)? If you cannot support it, I support it.

apdya commented on src/main/java/org/mybatis/spring/annotation/MapperScannerRegistrar.java:
```diff
@@ -126,6 +133,11 @@ void registerBeanDefinitions(AnnotationMetadata annoMeta, AnnotationAttributes a
 
+    AnnotationAttributes[] excludeFilterArray = annoAttrs.getAnnotationArray("excludeFilters");
+    for (AnnotationAttributes excludeFilters : excludeFilterArray) {
+      builder.addPropertyValue("excludeFilters", typeFiltersFor(excludeFilters));
```
thanks for pointing out that, I have modify my code and add combined filters test just now, I will commit later.

apdya commented:
> I added some comments.
> 
> @apdya
> 
> * Could you support this enhancement for the `<mybatis:scan>`(XML namespace feature)? If you cannot support it, I support it.
> * Please perform `./mvnw clean verify` (can apply code formatter and update license header)

thanks for review
I have modify my code as your comment, and updated this pr.
I would like to support XML namespace feature if this task is not urgent.
mvnw seems not work in my IDEA, i will figure it out soon. and I have add license header to all the new added file.

apdya commented:
> hi @kazuki43zoo 

> I have add some code to support exclude filter feature with xml. the new added element in xsd keep unified with Spring.

apdya commented:
still need some work

apdya commented on src/main/java/org/mybatis/spring/annotation/MapperScannerRegistrar.java:
```diff
@@ -145,6 +157,48 @@ void registerBeanDefinitions(AnnotationMetadata annoMeta, AnnotationAttributes a
+    String[] expressionArray = filterAttributes.getStringArray("pattern");
+    for (String expression : expressionArray) {
+      String rawName = filterType.toString();
+      if ("REGEX".equals(rawName)) {
```
new code has been change to switch-case

apdya commented on src/main/java/org/mybatis/spring/annotation/MapperScannerRegistrar.java:
```diff
@@ -145,6 +157,48 @@ void registerBeanDefinitions(AnnotationMetadata annoMeta, AnnotationAttributes a
+      }
+    }
+
+    String[] expressionArray = filterAttributes.getStringArray("pattern");
```
more test has been added

apdya commented:
> still need some work

> with latest commit, exclude-filter has been supported in <mybatis:scan> and test in `org.mybatis.spring.filter.xml.XmlScanFilterTest`

apdya commented:
@kazuki43zoo Is there anything else I need to change?

hazendaz commented:
@apdya Can you fix the merge conflict that exists now?

apdya commented:
> @apdya Can you fix the merge conflict that exists now?

@hazendaz happy to receive your reply, merge conflict has been resolved. whether this pr can be merged into the master branch now?

hazendaz commented:
@apdya I'll try to get to this in a week or two.  I believe all was good and just got missed.  I've got a reminder to review it again and take care of merging it in.  If you could, assuming there isn't one, open an issue that same support needs added to xml namespace as I see commented on here a few times so its not missed in future and at least helps others see that support is not present.

hazendaz commented:
as to a release as I'm sure that will come next after merging, probably end of month we could also get a release out.

apdya commented:
> @apdya I'll try to get to this in a week or two. I believe all was good and just got missed. I've got a reminder to review it again and take care of merging it in. If you could, assuming there isn't one, open an issue that same support needs added to xml namespace as I see commented on here a few times so its not missed in future and at least helps others see that support is not present.

 thanks for reply.   <mybatis:exclude-filter >(XML namespace feature) had been supported with commit [82ef68f] in this pr. test case: org.mybatis.spring.filter.xml.XmlScanFilterTest

hazendaz commented:
@apdya Merged.  I'll try to get a release towards end of the month.  If I miss that please do remind me here :)

hazendaz commented:
@apdya Sorry for delays.  I had a few personal issues that took up most my time last month.  So thank you for reminder.  I'll work to get this released next week.
